Feature/add bml cli commands to dynamic channel selection bml apis

adam1985d opened the pull request:
This PR adds CLI BML commands to the beerocks_cli for the following DCS BML APIS:
bml_set_dcs_continuous_scan_enable
bml_get_dcs_continuous_scan_enable
bml_set_dcs_continuous_scan_params
bml_get_dcs_continuous_scan_params
bml_start_dcs_single_scan
bml_get_dcs_scan_results

The commands naming are aligned to the called APIs without the 'bml_' initial:
set_dcs_continuous_scan_enable
get_dcs_continuous_scan_enable
set_dcs_continuous_scan_params
get_dcs_continuous_scan_params
start_dcs_single_scan
get_dcs_scan_results

The command's description is available in the beerocks_cli help menu

adam1985d commented:
![](https://github.trello.services/images/mini-trello-icon.png) [add cli to all DCS BMLs](https://trello.com/c/GrioG4Rj/80-add-cli-to-all-dcs-bmls)

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1661,3 +1661,49 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
 }
 #endif //BEEROCKS_RDKB
+
+int cli_bml::set_dcs_continuous_scan_enable(const std::string &radio_mac, int8_t enable)
```
why the `enable` is `int8` and not `bool`?

adam1985d replied:
the BML API uses int - and the CML CLI provides an int - so I don't see a value in int->bool->int conversion

mariomaz replied:
IMO, provided APIs should be always as clear as possible and thus enable should be bool in this case

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1661,3 +1661,49 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
+}
+
+int cli_bml::start_dcs_single_scan(const std::string &radio_mac, int dwell_time,
+                                   int channel_pool_size, const std::string &channel_pool)
```
I prefer using types that imply the number of bits that are used: uint32_t/int32_t.

apply this on all new stubs.

adam1985d replied:
done

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.h:
```diff
@@ -190,6 +190,18 @@ class cli_bml : public cli {
+                                       int dwell_time                  = BML_DCS_INVALID_PARAM,
+                                       int interval_time               = BML_DCS_INVALID_PARAM,
+                                       const std::string &channel_pool = std::string(),
+                                       int channel_pool_size           = BML_DCS_INVALID_PARAM);
```
I don't know think it is a good idea to define too many optional parameters, especially when they share common types.
I think a better idea will be to define a struct with default parameters and give it an argument.
(nitpick)

adam1985d replied:
might be, but will not be changed in this PR

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.h:
```diff
@@ -190,6 +190,18 @@ class cli_bml : public cli {
+                                       int dwell_time                  = BML_DCS_INVALID_PARAM,
+                                       int interval_time               = BML_DCS_INVALID_PARAM,
+                                       const std::string &channel_pool = std::string(),
+                                       int channel_pool_size           = BML_DCS_INVALID_PARAM);
```
There is no reason to add here `channel_pool_size` since it is a user interface. The function should automatically figure how many channels there are in the user input.

adam1985d replied:
discussed this with @tomereli  and he agrees with you. will be removed

tomereli replied:
Agree - the channel_pool should be a vector, you can parse it in the caller.

itayx replied:
Removed `channel_pool_size` throughout

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.h:
```diff
@@ -190,6 +190,18 @@ class cli_bml : public cli {
+    int set_dcs_continuous_scan_params(const std::string &radio_mac,
+                                       int dwell_time                  = BML_DCS_INVALID_PARAM,
+                                       int interval_time               = BML_DCS_INVALID_PARAM,
+                                       const std::string &channel_pool = std::string(),
```
Why are you passing `channel_pool` as a string and not a vector or array? The caller function should parse it and convert it to vector/array.

adam1985d replied:
none of the caller functions in the BML CLI perform any parsing. they are all a wrapper for getting the information from the CLI command to the cli_bml command.

tomereli replied:
I agree with Moran here - we should do the conversion in the BML CLI and get rid of the size of channel pool.

adam1985d replied:
fixed - removed the channel pool size from user input params. calculated from channel_pool input.

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1661,3 +1661,49 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
+}
+
+int cli_bml::get_dcs_scan_results(const std::string &radio_mac, unsigned int max_results_size,
+                                  int8_t is_single_scan)
```
why the` is_single_scan` is int8 and not bool?

adam1985d replied:
was changed to bool

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -507,6 +507,46 @@ void cli_bml::setFunctionsMapAndArray()
+                       1, STRING_ARG);
+    insertCommandToMap(
+        "bml_set_dcs_continuous_scan_params",
+        "<mac> <dwell_time_msec> [<interval_time>] [<channel_pool>] [<channel_pool_size>]",
```
You can define it like this only if you set all the optional arguments altogether. This is not the case here, so what you need to do is to explicitly write to what argument do you mean:
`<mac> <dwell_time_msec> [ interval_time=<interval_time>] [channel_pool=<channel_pool>]`

on the caller function you will need to parse it, and here convert all the argument types to `STRING_ARG`.

adam1985d replied:
ok - replaced the implementation to what you suggested (used the beerocks_cli_socket::client_beacon_11k_req as reference)

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1661,3 +1661,49 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
+
+int cli_bml::set_dcs_continuous_scan_enable(const std::string &radio_mac, int8_t enable)
+{
+    std::cout << "set_dcs_continuous_scan_enable" << std::endl;
```
Probably better to also print the radio mac and the enable.
Can we use `bool enable` instead of `int8_t`?

Same comments for the rest of the APIs

adam1985d replied:
1. bool instead of uint8_t:
- the BML API uses int - and the CML CLI provides an int - so I don't see a value in int->bool->int conversion
- in the 'get_dcs_scan_results()' I will move the int->bool translation to the caller function and update the signature
2. print the input values:
done

tomereli replied:
OK

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.h:
```diff
@@ -190,6 +190,18 @@ class cli_bml : public cli {
     int steering_client_measure(uint32_t steeringGroupIndex, const std::string &str_bssid,
                                 const std::string &str_client_mac);
 #endif
+    int set_dcs_continuous_scan_enable(const std::string &radio_mac, int8_t enable);
```
Please add doxygen to all new APIs either here or in the implementation file.
A rule of thumb - every new function comes with doxygen comments.

adam1985d replied:
I added inside the CPP file (not to clatter the class header)

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1123,6 +1123,67 @@ int cli_bml::bml_rdkb_steering_client_measure_caller(int numOfArgs)
+    } else if (numOfArgs == 3) {
+        return set_dcs_continuous_scan_params(args.stringArgs[0], args.intArgs[1], args.intArgs[2]);
+    } else if (numOfArgs == 4) {
+        std::cout << "Invalid number of arguments, set of channel_pool must also provide "
```
Why don't we print such an error in other failure cases?

adam1985d replied:
this is a special case where 2 params must be provided together or omitted together

adam1985d replied:
the code was removed as part of refactoring due to @morantr comment

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1806,5 +1913,63 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, unsigned int max
+        return -1;
+    }
+
+    struct BML_DCS_NEIGHBOR_AP *results = nullptr;
```
No need for `struct` keyword - its CPP.

adam1985d replied:
removed the struct keyword

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1806,5 +1913,63 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, unsigned int max
+
+    results =
+        (struct BML_DCS_NEIGHBOR_AP *)calloc(max_results_size, sizeof(struct BML_DCS_NEIGHBOR_AP));
+    if (results == nullptr) {
```
Did you run KW on this? please refrain from using `if (ptr == nullptr)` and instead use `if (!ptr)` to avoid KW issues.

adam1985d replied:
removed the <calloc> entirely - and will run KW

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1806,5 +1913,63 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, unsigned int max
+            std::cout << __func__ << "ERROR: results_count(" << results_count << ")"
+                      << " > max_results_size(" << max_results_size << ")" << std::endl;
+        } else if (status == 0 && results_count > 0) {
+            //todo: replace the temporary debug print with actual prints of the new results structure
```
I'm not sure I follow - what is temporary with the debug print? Aren't the results already in the new structure?

adam1985d replied:
no - since the PR that changes the structure is not available yet
(PRs 4/5): 
https://github.com/prplfoundation/prplMesh/pull/694

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1806,5 +1913,63 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, unsigned int max
+    struct BML_DCS_NEIGHBOR_AP *results = nullptr;
+
+    results =
+        (struct BML_DCS_NEIGHBOR_AP *)calloc(max_results_size, sizeof(struct BML_DCS_NEIGHBOR_AP));
```
There is absolutely no good reason to use `calloc` or `alloc` in c++, and especially in this function where you simply print the returned results and free them right after.
Simply declare the results on the stack:

`BML_DCS_NEIGHBOR_AP results[max_results_size];`

Also - why is the `max_results_size` argument even needed? This can also be statically declared in the function.

adam1985d replied:
the max_results_size is to simulate the max-results-size limit also used in <wave-api> in RDKB (currently set as 200)
when DCS supports ordering of the results by RSSI or another ranking, it is possible to the get only the X best results even if more are available

I replaced the <calloc> with a stack declaration.

tomereli replied:
The best would be to change the DCS WLAN APIs as well - is it feasible?

adam1985d replied:
it might be - you can create a JIRA to check this at a later time - shouldn't block this PR

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1767,13 +1767,27 @@ int cli_bml::set_dcs_continuous_scan_enable(const std::string &radio_mac, int8_t
+    int ret    = bml_get_dcs_continuous_scan_enable(ctx, radio_mac.c_str(), &enable);
+
+    if (ret == BML_RET_OK) {
+        std::cout << "dcs-continuous-scan-enable=" << ((enable == 1) ? "True" : "False")
```
string_utils::bool_str(enable);

adam1985d replied:
wasn't familiar with it - very useful - fixed, thanks

morantr commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1806,5 +1913,63 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, unsigned int max
+    struct BML_DCS_NEIGHBOR_AP *results = nullptr;
+
+    results =
+        (struct BML_DCS_NEIGHBOR_AP *)calloc(max_results_size, sizeof(struct BML_DCS_NEIGHBOR_AP));
```
Why didn't you declare it on the stack???
Anyway, this is C allocation. The CPP way is with `new`/`delete` and the better option will be to use smart pointers.

adam1985d replied:
@tomereli already mentioned it - I fixed it and am using a user-defined-sized array declared on the stack.

mariomaz commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -507,6 +507,46 @@ void cli_bml::setFunctionsMapAndArray()
+        "<mac> <dwell_time_msec> [<interval_time>] [<channel_pool>] [<channel_pool_size>]",
+        "Set the continuous scan params for the given AP mac: dwell_time - dwell time in "
+        "miliseconds, interval_time - interval time in seconds,"
+        " channel_pool - channels seperated by commas (set of channel_pool must also provide "
```
typo: should be "separated"

adam1985d replied:
fixed - thanks

mariomaz commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -507,6 +507,46 @@ void cli_bml::setFunctionsMapAndArray()
+        "bml_set_dcs_continuous_scan_params",
+        "<mac> <dwell_time_msec> [<interval_time>] [<channel_pool>] [<channel_pool_size>]",
+        "Set the continuous scan params for the given AP mac: dwell_time - dwell time in "
+        "miliseconds, interval_time - interval time in seconds,"
```
typo: "milliseconds"
This typo is repeated in other commands

adam1985d replied:
fixed - thanks

mariomaz commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -507,6 +507,46 @@ void cli_bml::setFunctionsMapAndArray()
+                       static_cast<pFunction>(&cli_bml::get_dcs_continuous_scan_params_caller), 1,
+                       1, STRING_ARG);
+    insertCommandToMap(
+        "bml_start_dcs_single_scan", "<mac> <dwell_time_msec> <channel_pool_size> <channel_pool>",
```
For the sake of consistency, channel_pool_size and channel_pool should be given in the same order in all commands.

adam1985d replied:
I felt the same itch - but this was aligned to the signature of DCS BML APIS.
we decided to remove the channel_pool_size completely (and parse it from the string of channel_pool) - so will be fixed

adam1985d replied:
update - removed the channel_pool_size from user input

mariomaz commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1661,3 +1762,214 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
+
+int cli_bml::get_dcs_continuous_scan_params(const std::string &radio_mac)
+{
+    std::cout << "get_dcs_continuous_scan_params" << std::endl;
```
You should probably be using __func__ instead of a literal string to avoid copy & paste errors. This way maintenance is easier (if method name changes, there is no need to change anything else). 
This is something to be considered in all the methods in this file.

adam1985d replied:
fixed

mariomaz commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1661,3 +1762,214 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
+                                                 &interval_time, channel_pool, &channel_pool_size);
+
+    if (ret == BML_RET_OK) {
+        std::cout << "dcs-continuous-scan-params for mac:" << std::endl
```
MAC address is not being printed

adam1985d replied:
fixed

mariomaz commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1661,3 +1762,214 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
+    std::cout << "start_dcs_single_scan" << std::endl;
+
+    if (dwell_time <= 0) {
+        std::cout << __func__ << "invalid input: dwell_time(" << dwell_time << ") <= 0"
```
Method name is being printed together with next literal string

adam1985d replied:
fixed

mariomaz commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1661,3 +1762,214 @@ int cli_bml::steering_client_measure(uint32_t steeringGroupIndex, const std::str
+        return -1;
+    }
+
+    if (channels_size > 0 && channels_size != (size_t)channel_pool_size) {
```
IMO, parentheses in conditionals, even if they are not strictly required, help writing code that is clear and correct

adam1985d replied:
this is not maintained across the code - but if you find it more readable, I don't mind adding it

fixed.

mariomaz replied:
This was not something to be addressed in this PR but to be discussed if code writing rules are ever revisited.

mariomaz commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -1123,6 +1163,67 @@ int cli_bml::bml_rdkb_steering_client_measure_caller(int numOfArgs)
+
+int cli_bml::get_dcs_continuous_scan_enable_caller(int numOfArgs)
+{
+    if (numOfArgs == 1) {
```
I've been using Yoda conditions for a long time now that these lines make me feel uncomfortable. 
Some static code analyzers would also complain.

adam1985d replied:
I am familiar with this approach - but this is not maintained in the prpl.

if @tomereli and @arnout  want to add this requirement to prplmesh guidelines it is OK, but a short search of the following regex. <[a-z,A-Z] == [0-9]> produced 375 results in 93 files

will not be handled in this PR, but we can set a task for this change across the code if decided

tomereli replied:
Fix this, we will.
![image](https://user-images.githubusercontent.com/13386849/72826517-ab438c80-3c81-11ea-8b3e-4b94d744280f.png)

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -24,6 +24,17 @@ using namespace net;
 /////////////////////////// Local Module Functions ///////////////////////////
 //////////////////////////////////////////////////////////////////////////////
 
+static const std::string string_from_int_array(void *arr, size_t arr_max_size)
```
The name of this function implies it creates a string from an array of integers. So it should be 
```
static const std::string string_from_int_array(int arr[], size_t len);
```
Also - please add a doxygen comment explaining this function.

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2021,28 +2032,40 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+                          << "  channel=" << int(res.ap_Channel) << std::endl
+                          << "  signal_strength=" << int(res.ap_SignalStrength) << std::endl
+                          << "  security_mode_enabled="
+                          << string_from_int_array(res.ap_SecurityModeEnabled,
```
I'm think that security mode should be a single value - I'm dazzled why you chose array. Please make it a single value unless there's a very good reason why you need an array. Then we can get rid of this weird `string_from_int_array` function.

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2021,28 +2032,40 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
+                                                   BML_CHANNEL_SCAN_ENUM_LIST_SIZE)
+                          << std::endl
+                          << "  encryption_mode="
+                          << string_from_int_array(res.ap_EncryptionMode,
```
Same here - there is a single encryption mode set at a given time. I don't understand why it is an array.

tomereli commented on controller/src/beerocks/cli/beerocks_cli_bml.cpp:
```diff
@@ -2021,28 +2032,40 @@ int cli_bml::get_dcs_scan_results(const std::string &radio_mac, uint32_t max_res
-            //todo: replace the temporary debug print with actual prints of the new results structure
             for (size_t i = 0; i < results_count; i++) {
                 auto res = results[i];
                 std::cout << "result[" << i << "]:" << std::endl
```
Instead of this whole big print, why not add `inline std::ostream &operator<<(std::ostream &os, const BML_NEIGHBOR_AP &neigh);` to `struct BML_NEIGHBOR_AP` in `bml_defs.h` then you can replace this with:
```
for (size_t i = 0; i < results_count; i++) {
    std::cout << "result[" << i << "]:" << std::endl << results[i] << std::endl;
}
```

tomereli reviewed (approved):
I added some more comments, I think you should fix them but if this is urgent it can be done as part of #665 , so approving again.

tomereli commented:
Finally, please rebase and squash before merge.

adam1985d committed:
cli_bml:add BML CLI commands stubs to DCS BML APIs

added the following stubs of BML CLI commands of DCS BML APIs:
set_dcs_continuous_scan_enable
get_dcs_continuous_scan_enable
set_dcs_continuous_scan_params
get_dcs_continuous_scan_params
start_dcs_single_scan
get_dcs_scan_results

Signed-off-by: Adam Dov <adamx.dov@intel.com>
Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

adam1985d committed:
cli_bml:add caller functions to BML CLI commands

Signed-off-by: Adam Dov <adamx.dov@intel.com>
Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

adam1985d committed:
cli_bml:setFunctionsMapAndArray() - add DCS BML commands

The insertion of the BML commands is performed in the
setFunctionsMapAndArray() using the insertCommandToMap command.

After this change, the following commands are available in the BML CLI:
set_dcs_continuous_scan_enable
get_dcs_continuous_scan_enable
set_dcs_continuous_scan_params
get_dcs_continuous_scan_params
start_dcs_single_scan
get_dcs_scan_results

Signed-off-by: Adam Dov <adamx.dov@intel.com>
Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

adam1985d committed:
cli_bml:implement the DCS BML CLI commands

The implementation of each command is aligned with the existing
DCS wave functions implementation in "wave-api" in RDKB.
A detailed description of each command use is available in the
CLI help menu.

Signed-off-by: Adam Dov <adamx.dov@intel.com>
Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

adam1985d commented:
fixed all code review comments and passed all tests
